### ISSUE TITLE
Compatablilty with other clones/vendors

### ISSUE THREAD

**Issue by NF3RN0** [pixeltris/GK6X#5]
I have recently purchased this [similar clone](https://www.amazon.com/Mechanical-Keyboard-Waterproof-Computer-Programmable/dp/B07W5PD6VS) from Amazon. Seems to me that they use _almost exact_ copy of hardware and software, possibly just rebranding the entire thing. Upon opening the device up I have noticed that they did at least source the outside shell from Jikedingzhi as it is stamped with their logo.

It is important to note that the original software from the Jikedingzhi website does not work with my device, I have already tried that. 

The driver did successfully build in Linux. Then when I ran the app I did receive errors that it was unable to perform a handshake with the device.

`[15:33:07.2895780] Handshake failed (USB-HID Gaming Keyboard) Couldn't find data for keyboard`

**$ lsusb**

`Bus 001 Device 008: ID 1ea7:0907 SHARKOON Technologies GmbH USB-HID Gaming Keyboard`

I cannot expect you to put in the work for supporting this device but I was wondering if you might have any insight on what I could do to get this working. I feel like my device is not too far off from compatibility.

Cheers.

**Comment by pixeltris** [comment]
Based on that log message it does sound like the keyboard uses the same protocol as it gets a decent way into the handshake before printing that message. What that message means is that there isn't any keyboard layout info defined in the data files. The data files were obtained from the official software (which is probably why that keyboard doesn't work on the official software either, as the data files don't exist). You need to do the following:

- Place a breakdown down on this line and get the model id / firmware id (if it isn't printed out already) https://github.com/pixeltris/GK6X/blob/81761765f020962ed04ad68f8f0837facda7a1fa/GK6X/KeyboardDeviceManager.cs#L209
- Add that info to https://github.com/pixeltris/GK6X/blob/master/Build/Data/device/modellist.json
- Create a folder here with that model id as the folder name https://github.com/pixeltris/GK6X/tree/master/Build/Data/device
- Copy files from a keyboard with a similar layout into your folder (e.g. folder 655491085 which is one of the GK61 boards). See if that works and if not try copying the data from another folder until it looks good, or manually edit the files.

**Comment by NF3RN0** [comment]
Thank you for the detailed information. I got it to work.

Here is the printout from your provided line of code.

`modelID: 656801811 FWID: 34620929`

That should have been easier but I realized that I was using the wrong firmware Id, I had to convert the printed ID to hex. I dug a little bit deeper into the software provided by Kemove. The device data along with the model list description is included with their application. So it was a matter of putting everything in the right place. If you want I could provide the additions with a pull request.

Here is the working result after I used the provided device information.

`Connected to device 'DK61(wired)' model:656801811 fw:v1.20`

I have tested the functionality with sample.txt as well as making my own personal keyboard settings. Everything seems to just work.


Thanks again!
